platforms/ide/tooling-api/src/crossVersionTest/groovy/org/gradle/integtests/tooling/r812/CustomTestEventsCrossVersionSpec.groovy

```diff
@@ -0,0 +1,216 @@
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.gradle.integtests.tooling.r812
+
+import groovy.transform.CompileStatic
+import org.gradle.integtests.tooling.TestLauncherSpec
+import org.gradle.integtests.tooling.fixture.TargetGradleVersion
+import org.gradle.integtests.tooling.fixture.ToolingApiVersion
+import org.gradle.tooling.TestLauncher
+import org.gradle.tooling.events.OperationDescriptor
+import org.gradle.tooling.events.task.TaskOperationDescriptor
+import org.gradle.tooling.events.test.TestOperationDescriptor
+
+@ToolingApiVersion('>=8.8')
+@TargetGradleVersion(">=8.12")
+class CustomTestEventsCrossVersionSpec extends TestLauncherSpec {
+    @Override
+    void addDefaultTests() {
+    }
+
+    @Override
+    String simpleJavaProject() {
+        ""
+    }
+
+    def "reports custom test events"() {
+        given:
+        if (!buildFile.delete()) {
+            throw new RuntimeException("Could not delete build file")
+        }
+        buildFile("""
+            import java.time.Instant
+
+            abstract class CustomTestTask extends DefaultTask {
+                @Inject
+                abstract TestEventService getTestEventService()
+
+                @TaskAction
+                void runTests() {
+                   try (def generator = getTestEventService().generateTestEvents("Custom test root")) {
+                       generator.started(Instant.now())
+                       try (def mySuite = generator.createCompositeNested("My Suite")) {
+                            mySuite.started(Instant.now())
+                            try (def myTest = mySuite.createSolitaryNested("MyTestInternal", "My test!")) {
+                                 myTest.started(Instant.now())
+                                 myTest.output(TestOutputEvent.Destination.StdOut, "This is a test output on stdout")
+                                 myTest.output(TestOutputEvent.Destination.StdErr, "This is a test output on stderr")
+                                 myTest.completed(Instant.now(), TestResult.ResultType.SUCCESS)
+                            }
+                            mySuite.completed(Instant.now(), TestResult.ResultType.SUCCESS)
+                       }
+                       generator.completed(Instant.now(), TestResult.ResultType.SUCCESS)
+                   }
+                }
+            }
+
+            tasks.register("customTest", CustomTestTask)
+        """)
+
+        when:
+        launchTests { TestLauncher launcher ->
+            launcher.withTaskAndTestClasses(':customTest', ['no idea what to do with this yet'])
```

Comment:
We don't need to make this work, it would be OK to use BuildLauncher.

Can we still receive test events if we do?